### ISSUE TITLE
Invalid suppression in `TypeConverter` when loading a custom converter type

### ISSUE THREAD

**Issue by vitek-karas** [dotnet/runtime#95265]
This suppression is invalid:
https://github.com/dotnet/runtime/blob/362854418420e2937f8bcd11ca45664d8ba2563a/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.ReflectedTypeData.cs#L487-L488

It relies on the behavior like
```csharp
namespace ns {
    public class MyConverter : TypeConverter {}
}

[TypeConverter("ns.MyConverter")]
public class MyType : ILibInterface {}
```

The above will work because the simple type name in the attribute refers to the type in the same assembly as the type which has the attribute applied to it.

The problem is that the same is not true for case where the attribute comes from a base class or interface. For example if the interface is defined as below in a separate assembly `Lib`:
```csharp
namespace ns {
    public class MyConverter : TypeConverter {}
}

[TypeConverter("ns.MyConverter")]
public interface ILibInterface {}
```

Now remove the `TypeConverter` attribute from the `MyType` in the app like this, but keep the custom converter type:
```csharp
namespace ns {
    public class MyConverter : TypeConverter {}
}

public class MyType : ILibInterface {}
```

And now use it:
```csharp
TypeDescriptor.GetConverter(typeof(MyType)).GetType().FullName
```

This will print out `ns.MyConverter, app` - so it will load the custom converter from the app! Not the one from the Lib where the attribute is used. This is because of how the type lookup is implemented in https://github.com/dotnet/runtime/blob/362854418420e2937f8bcd11ca45664d8ba2563a/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.ReflectedTypeData.cs#L506 where it uses the assembly of the type on which we're getting the converter.

Now in trimming: `TypeConverterAttribute.ctor([DynamicallyAccessedMembers(PublicConstructors)] string)`.
The behavior of this is defined as resolving the string passed in in the context of the scope where the call is made. So in the case of the attribute used on the interface above, the string is resolved to a type in the context of the `ILibInterface`, so it resolves the custom converter from the `Lib` assembly. The trimmer will preserve that, but it will remove the custom converter type from the `App` since there are no references to it.
At runtime the algorithm to lookup the type name will fail to resolve the type name in the attribute and fallback to the default converter instead.

I can't think of a solution to this, because the trimmer has no idea what type of resolution logic is used, it warns, but the warning is suppressed.

Important note - currently `TypeDescriptor.GetConverter` is annotated with `RequiresUnreferecedCode` so the above will produce a trimmer warning. So, the bug here is effectively covered by that. But the explanation of why we need the RUC in this case is at least partially incorrect. And efforts to optimize the annotations in the BCL library will be affected by the invalid suppression regardless.

/cc @sbomer @eerhardt @LakshanF

**Comment by steveharter** [comment]
@vitek-karas were you going to further triage\research this as part of linkability?

**Comment by vitek-karas** [comment]
Currently I don't have any plans to tackle this. I ran into it during some investigation, so I created the bug.

**Comment by buyaa-n** [comment]
@sbomer is this still planned for V9.0?

**Comment by ericstj** [comment]
@sbomer does this need to be fixed for 9.0?  It looks like the mentioned suppression has been there since 6.0.  cc @eerhardt

**Comment by steveharter** [comment]
> Trimming didn't keep ExpandableObjectConverter because it can't see through the Type.GetType of a substring of the input.

So when using an assembly name ("assemblyDoesntExist.dll") for the assembly type name like this:
```
[TypeConverter ("System.ComponentModel.ExpandableObjectConverter, assemblyDoesntExist.dll")]
```
where the assembly name is later [removed at run-time](https://github.com/dotnet/runtime/blame/9d7551ecd3e1b1b6f224eabecc1c7f91a41f259c/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.ReflectedTypeData.cs#L535-L546) before the Type.GetType lookup, and the specified assembly didn't actually exist (before trimming), then the trimmer won't know what to do since the assembly never existed?

If that's the case then I agree with this:

> The example is a bit contrived. I don't think this is likely to cause real world problems so probably not important to get fixed in 9.0, but I'd let area owners decide.

**Comment by steveharter** [comment]
I'm going to close this; if it causes a real-world issue then we can re-evaluate.

The given type name should not have an invalid assembly name, perhaps and older one to support design-time cases, but not invalid. The name is stripped out due to the comment in the source due to design-time:
``` cs
// At design time, it's possible for us to reuse
// an assembly but add new types. The app domain
// will cache the assembly based on identity, however,
// so it could be looking in the previous version
// of the assembly and not finding the type. We work
// around this by looking for the non-assembly qualified
// name, which causes the domain to raise a type
// resolve event.
```

Perhaps that code could throw if an assembly name is specified?
